source/components/ui/msm-click-to-edit/msm-click-to-edit.directive.js

```diff
@@ -0,0 +1,126 @@
+/** Based on https://github.com/GabiGrin/angular-editable-text */
+(function () {
+  'use strict';
+
+  /**
+   * @ngdoc directive
+   * @name components.ui.msmEditableText
+   * @restrict 'E'
+   *
+   * @description Renders an editable text input
+   *
+   * @param {string} msmEditableText The model to be edited
+   * @param {string} editMode The edit mode
+   * @param {string} placeholder The placeholder for the model if empty
+   * @param {function} onChange A callback function on model change
+   */
+  angular.module('msm.components.ui')
+      .directive('msmEditableText', EditableText);
+
+  function EditableText() {
+    return {
+      scope: {
+        msmEditableText: '=',
+        editMode: '=',
+        placeholder: '@',
+        onChange: '&'
+      },
+      transclude: true,
+      templateUrl: 'components/ui/msm-click-to-edit/msm-click-to-edit.html',
+      link: function (scope, elem, attrs) {
+        var input = elem.find('input');
+        var lastValue;
+        scope.saveValue = true;
+        scope.isEditing = !!scope.editMode;
+        scope.editingValue = scope.msmEditableText;
+
+        scope.onBlur = function() {
+          scope.isEditing = false;
+        };
+
+        scope.onKeyPress = function(event) {
+          if (event) {
+            // enter
+            if (event.which === 13) {
+              scope.isEditing = false;
+            }
+            // escape
+            if (event.which === 27) {
+              scope.saveValue = false;
+              scope.isEditing = false;
+            }
+          }
+        };
+
+        scope.onClick = function() {
+          scope.isEditing = true;
+        };
+
+        scope.onFocus = function() {
+          scope.isEditing = true;
+        };
+
+        elem.addClass('editable-text');
+
+        scope.$watch('isEditing',
+            function (val, oldVal) {
+              if (attrs.editMode !== undefined) {
+                scope.editMode = val;
+              }
+              if (val) {
+                elem['addClass']('editing');
+                var inputElm = input[0];
```

Comment:
Thank you! In javascript it is considered really bad practice to create variables this way. It can lead to errors and confusion as there is no block scope like I said. This can be avoided by using named functions in if and else statements (it also improves readability and development speed a lot).